File: app/meetings.py
"""定例会議。ガントの 1 行に、開催日を点で並べるためのもの。

会議を 1 回ずつタスクにすると、週次の定例だけで 1 年に 52 行になり、
ガントが会議で埋まってしまう。ここでは「毎週火曜」のような決まりだけを持ち、
開催日は見る期間のぶんだけその場で数える。先の予定まで出るのはそのため。

1 回ごとの中止・日にち変更は、例外として別に持つ。例外の鍵は「その回の
予定日」（休日でずらしたあとの日）で、画面で点を押した日と一致する。
決まりを後から変えて予定日が消えた例外は、ただ効かなくなるだけにしてある。
"""
from datetime import date, timedelta

from . import db, holidays

FREQ_LABEL = {"weekly": "毎週", "monthly": "毎月"}
WEEKDAY_LABEL = "月火水木金土日"
HOLIDAY_RULES = {
    "skip": "その回は休み",
    "next": "翌営業日にずらす",
    "prev": "前営業日にずらす",
    "keep": "そのまま",
}
# 休日でずらすときに、どこまで先（前）を探すか
SHIFT_LIMIT = 14
# 一度に数える期間の上限。全体ガントで数年分を並べても重くならない程度
MAX_SPAN_DAYS = 1500


def parse_weekdays(value):
    out = set()
    for part in str(value or "").split(","):
        part = part.strip()
        if part.isdigit() and 0 <= int(part) <= 6:
            out.add(int(part))
    return sorted(out)


def _month_index(day):
    return day.year * 12 + day.month - 1


def _last_day(year, month):
    nxt = date(year + month // 12, month % 12 + 1, 1)
    return (nxt - timedelta(days=1)).day


def _nth_weekday(year, month, nth, weekday):
    """その月の第 nth weekday。nth が -1 なら最終。無い月（第5など）は None。"""
    if nth == -1:
        last = date(year, month, _last_day(year, month))
        return last - timedelta(days=(last.weekday() - weekday) % 7)
    first = date(year, month, 1)
    day = first + timedelta(days=(weekday - first.weekday()) % 7 + 7 * (nth - 1))
    return day if day.month == month else None


def raw_dates(m, lo, hi):
    """休日を考えない、決まりどおりの開催日。lo〜hi と開始日〜終了日の両方に入るもの。"""
    start = m["start_on"]
    lo = max(lo, start)
    if m.get("end_on"):
        hi = min(hi, m["end_on"])
    if hi < lo:
        return []
    interval = max(1, int(m.get("interval_n") or 1))
    out = []
    if m["freq"] == "weekly":
        weekdays = parse_weekdays(m.get("weekdays")) or [start.weekday()]
        anchor = start - timedelta(days=start.weekday())
        day = lo
        while day <= hi:
            if day.weekday() in weekdays:
                weeks = ((day - timedelta(days=day.weekday())) - anchor).days // 7
                if weeks % interval == 0:
                    out.append(day)
            day += timedelta(days=1)
        return out
    # monthly
    base = _month_index(start)
    for index in range(_month_index(lo), _month_index(hi) + 1):
        if (index - base) % interval:
            continue
        year, month = divmod(index, 12)
        month += 1
        if m.get("month_mode") == "nth":
            day = _nth_weekday(year, month, int(m.get("nth") or 1),
                               int(m.get("nth_weekday") or 0))
        else:
            wanted = int(m.get("month_day") or start.day)
            day = date(year, month, min(wanted, _last_day(year, month)))
        if day and lo <= day <= hi:
            out.append(day)
    return out


def _shift(day, rule, is_off):
    """休日に当たった回の扱い。休みにするなら None。"""
    if rule == "keep" or not is_off(day):
        return day
    if rule == "skip":
        return None
    step = timedelta(days=1 if rule == "next" else -1)
    for _ in range(SHIFT_LIMIT):
        day += step
        if not is_off(day):
            return day
    return None


def scheduled(m, lo, hi, is_off):
    """休日の扱いまで済ませた予定日。[(決まりどおりの日, 予定日)] を日付順で。

    ずらした結果が期間に入ってくる回もあるので、少し広めに数えてから絞る。
    ずらした先が別の回と重なったら 1 回にまとめる。
    """
    pad = timedelta(days=SHIFT_LIMIT + 1)
    seen = {}
    for planned in raw_dates(m, lo - pad, hi + pad):
        day = _shift(planned, m.get("holiday_rule") or "next", is_off)
        if day and lo <= day <= hi and day not in seen:
            seen[day] = planned
    return sorted(((planned, day) for day, planned in seen.items()), key=lambda p: p[1])


def is_scheduled(m, day, is_off):
    return any(d == day for _, d in scheduled(m, day, day, is_off))


def occurrences(m, lo, hi, is_off, exceptions):
    """lo〜hi に並べる回。中止した回も「中止」として返す（消えると気づけないため）。

    exceptions は {予定日: 例外の行}。日にちを変えた回は、変えた先の日に出す。
    """
    out = []
    for planned, day in scheduled(m, lo, hi, is_off):
        ex = exceptions.get(day)
        if ex and ex["action"] == "move":
            continue                      # 変えた先のほうで出す
        out.append({
            "date": day.isoformat(), "planned": day.isoformat(),
            "shifted_from": planned.isoformat() if planned != day else None,
            "status": "cancelled" if ex else "normal",
            "note": ex["note"] if ex else "",
        })
    for day, ex in exceptions.items():
        moved = ex.get("moved_to")
        if ex["action"] != "move" or not moved or not lo <= moved <= hi:
            continue
        # 決まりを変えて元の回が無くなった例外は、もう効かせない
        if not is_scheduled(m, day, is_off):
            continue
        out.append({
            "date": moved.isoformat(), "planned": day.isoformat(),
            "shifted_from": None, "status": "moved", "note": ex["note"],
        })
    out.sort(key=lambda o: (o["date"], o["planned"]))
    return out


def off_days(lo, hi):
    """lo〜hi（前後に余裕を持たせて）の休日判定。問い合わせは 1 回で済ませる。"""
    pad = timedelta(days=SHIFT_LIMIT + 2)
    # 祝日を使わない設定なら、土日だけを休みとして扱う（ガントの網掛けと揃える）
    extra = (holidays.holidays_between(lo - pad, hi + pad)
             if db.get_setting("use_holidays", "1") == "1" else {})
    return lambda day: holidays.is_holiday(day, extra)


def load_exceptions(meeting_ids):
    """{会議 id: {予定日: 例外}}。"""
    out = {}
    if not meeting_ids:
        return out
    for row in db.query(
            "SELECT meeting_id, on_date, action, moved_to, note FROM meeting_exceptions "
            "WHERE meeting_id IN %s", (tuple(meeting_ids),)):
        out.setdefault(row["meeting_id"], {})[row["on_date"]] = row
    return out


def describe(m):
    """「隔週 火・木 10:00」のような短い説明。"""
    interval = max(1, int(m.get("interval_n") or 1))
    if m["freq"] == "weekly":
        head = "毎週" if interval == 1 else ("隔週" if interval == 2 else
                                            "{}週ごと".format(interval))
        days = "・".join(WEEKDAY_LABEL[d] for d in parse_weekdays(m.get("weekdays")))
        text = "{} {}".format(head, days).strip()
    else:
        head = "毎月" if interval == 1 else "{}か月ごと".format(interval)
        if m.get("month_mode") == "nth":
            nth = int(m.get("nth") or 1)
            which = "最終" if nth == -1 else "第{}".format(nth)
            text = "{} {}{}曜".format(head, which, WEEKDAY_LABEL[int(m.get("nth_weekday") or 0)])
        else:
            text = "{} {}日".format(head, int(m.get("month_day") or 1))
    if m.get("time_text"):
        text += " " + m["time_text"]
    return text

File: static/js/views/meetings.js
/* 定例会議。ガントの 1 行に開催日を点で並べる。
 *
 * ここにあるのは入力まわりだけ（決まりの登録と、1 回ごとの中止・日にち変更）。
 * 開催日の計算はサーバーが持っている。祝日の暦や例外と突き合わせるため、
 * 画面側で同じ計算をもう 1 つ持つと、どこかで食い違うので持たない。 */
import { api } from '../api.js';
import {
  confirmDialog, el, fill, formatDate, openModal, parseDate, toast, today, toISO,
} from '../util.js';
import { markerChar, store } from '../store.js';
import { option } from './pickers.js';
import { buildTree } from './tasks.js';

const WEEKDAYS = ['月', '火', '水', '木', '金', '土', '日'];
const WEEK_INTERVALS = [[1, '毎週'], [2, '隔週'], [3, '3週ごと'], [4, '4週ごと']];
const MONTH_INTERVALS = [[1, '毎月'], [2, '2か月ごと'], [3, '3か月ごと'], [6, '半年ごと']];
const NTHS = [[1, '第1'], [2, '第2'], [3, '第3'], [4, '第4'], [-1, '最終']];
const HOLIDAY_RULES = [
  ['skip', 'その回は休み'],
  ['next', '翌営業日にずらす'],
  ['prev', '前営業日にずらす'],
  ['keep', 'そのまま'],
];
// 週次の定例は祝日なら休み、月次の締め会議はずらす、というのが多い
const DEFAULT_HOLIDAY = { weekly: 'skip', monthly: 'next' };

/** 「9/23(火)」 */
export function dayLabel(iso) {
  return formatDate(iso, true);
}

/**
 * ガントで置く場所の候補。タスクの木の並びのまま、見出しとマイルストーンも含めて出す。
 * 見出しを選ぶと、その区切りの先頭に並ぶ。マイルストーンを選ぶと、その下に並ぶ
 * （「リリース判定会」のように、節目にひもづく打ち合わせがあるため）。
 */
export function placementCandidates(tasks) {
  const { children } = buildTree(tasks);
  const out = [];
  const walk = (parentId, depth) => {
    for (const task of children.get(parentId) || []) {
      let label = task.title;
      if (task.is_heading) label = `【見出し】${task.title}`;
      else if (task.is_milestone) label = `${markerChar(task)} ${task.title}`;
      out.push({ id: task.id, depth, heading: Boolean(task.is_heading), label });
      walk(task.id, depth + 1);
    }
  };
  walk(null, 0);
  return out;
}

/** 月曜=0 の曜日番号。Date.getDay() は日曜=0 なので直す。 */
function mondayIndex(date) {
  return (date.getDay() + 6) % 7;
}

/**
 * 定例の登録・編集。保存したら true、消したら 'deleted' を返す。
 * @param {object} project 追加先（編集のときは meeting.project_id を使う）
 * @param {object|null} meeting 編集する定例
 * @param {Array} tasks 置き場所の候補（そのプロジェクトのタスク）
 */
export async function openMeetingForm({ project, meeting = null, tasks = [] }) {
  const now = today();
  const state = {
    freq: meeting?.freq || 'weekly',
    weekdays: new Set(meeting?.weekdays?.length ? meeting.weekdays : [mondayIndex(now)]),
    monthMode: meeting?.month_mode || 'day',
    interval: Number(meeting?.interval_n || 1),
    dates: new Set(meeting?.dates || []),
    // 休日の扱いを自分で選んだら、繰り返しの種類を変えても上書きしない
    holidayTouched: Boolean(meeting),
  };
  const f = {};
  const freqHost = el('div', {});
  const preview = el('div', { class: 'hint meeting-preview' });
  // 「日付を指定」では使わない欄。選んだ日は休日でもずらさず、開始・終了も要らない
  const ruleOnly = [];
  let previewTimer = null;
  let previewSeq = 0;

  const payload = () => ({
    parent_id: f.parent.value ? Number(f.parent.value) : null,
    title: f.title.value.trim(),
    freq: state.freq,
    interval_n: Number(f.interval?.value || 1),
    weekdays: [...state.weekdays].sort(),
    month_mode: state.monthMode,
    month_day: f.monthDay ? Number(f.monthDay.value) || null : null,
    nth: f.nth ? Number(f.nth.value) : null,
    nth_weekday: f.nthWeekday ? Number(f.nthWeekday.value) : null,
    dates: [...state.dates].sort(),
    time_text: f.time.value.trim(),
    holiday_rule: f.holiday.value,
    start_on: f.start.value || null,
    end_on: f.end.value || null,
  });

  /** 入力に合わせて「次は 9/29(火)・10/6(火)…」を出す。 */
  const drawPreview = () => {
    clearTimeout(previewTimer);
    if (state.freq === 'dates' && !state.dates.size) {
      // まだ 1 日も選んでいないうちは、叱らずに空けておく
      previewSeq += 1;
      preview.textContent = '';
      return;
    }
    previewTimer = setTimeout(async () => {
      const seq = ++previewSeq;
      try {
        const data = await api.post('/api/meetings/preview', payload());
        if (seq !== previewSeq) return;             // 古い応答は捨てる
        preview.classList.remove('error');
        const none = state.freq === 'dates'
          ? '選んだ日はすべて過ぎています'
          : 'この決まりでは、この先 1 年ほど開催日がありません';
        preview.textContent = data.next.length
          ? `次の開催: ${data.next.map((n) => dayLabel(n.date)
            + (n.shifted_from ? `（${formatDate(n.shifted_from)}が休日のため）` : '')).join('、')}`
          : none;
      } catch (error) {
        if (seq !== previewSeq) return;
        preview.classList.add('error');
        preview.textContent = error.message;
      }
    }, 250);
  };

  /** 選んだ日の一覧。押すと外せる。 */
  const drawDates = (host) => {
    const sorted = [...state.dates].sort();
    fill(host, ...(sorted.length
      ? sorted.map((iso) => el('button', {
        type: 'button', class: 'date-chip', title: 'この日を外す',
        onClick: () => { state.dates.delete(iso); drawDates(host); drawPreview(); },
      }, el('span', { text: dayLabel(iso) }), el('span', { class: 'x', text: '×' })))
      : [el('span', { class: 'hint', text: 'まだ日付がありません' })]));
  };

  const drawFreq = () => {
    f.monthDay = null;
    f.nth = null;
    f.nthWeekday = null;
    for (const node of ruleOnly) node.hidden = state.freq === 'dates';
    if (state.freq === 'dates') {
      const chips = el('div', { class: 'date-chips' });
      const picker = el('input', { class: 'input', type: 'date', style: { maxWidth: '170px' } });
      const add = () => {
        if (!picker.value) { toast('日付を選んでください', 'error'); return; }
        state.dates.add(picker.value);
        picker.value = '';
        drawDates(chips);
        drawPreview();
      };
      // 日付を選んだらすぐ足す。何日も続けて選べるように、欄は空に戻す
      picker.addEventListener('change', () => { if (picker.value) add(); });
      drawDates(chips);
      fill(freqHost, el('div', { class: 'field' },
        el('label', { text: '開催日（何日でも）' }),
        el('div', { style: { display: 'flex', gap: '8px', alignItems: 'center' } },
          picker, el('span', { class: 'hint', text: '選ぶとその場で足されます' })),
        chips,
        el('div', { class: 'hint',
          text: '1 回だけの会議も、不定期に何回かある会議も、1 行にまとめて並べられます。' })));
      return;
    }
    const intervals = state.freq === 'weekly' ? WEEK_INTERVALS : MONTH_INTERVALS;
    const current = state.interval;
    f.interval = el('select', {
      class: 'select',
      onChange: () => { state.interval = Number(f.interval.value); drawPreview(); },
    },
      ...intervals.map(([value, label]) => option(value, label, value === current)));
    if (!intervals.some(([value]) => value === current)) {
      // 画面の選択肢に無い間隔（5週ごとなど）で保存されていても消さない
      f.interval.appendChild(option(current, `${current}${state.freq === 'weekly' ? '週' : 'か月'}ごと`, true));
    }

    if (state.freq === 'weekly') {
      fill(freqHost, el('div', { class: 'row' },
        el('div', { class: 'field', style: { flex: '0 0 130px' } },
          el('label', { text: '間隔' }), f.interval),
        el('div', { class: 'field' },
          el('label', { text: '曜日' }),
          el('div', { class: 'wd-row' }, ...WEEKDAYS.map((label, index) => el('button', {
            type: 'button', class: `wd-btn${state.weekdays.has(index) ? ' active' : ''}`,
            onClick: (event) => {
              if (state.weekdays.has(index)) state.weekdays.delete(index);
              else state.weekdays.add(index);
              event.currentTarget.classList.toggle('active', state.weekdays.has(index));
              drawPreview();
            },
          }, label))))));
      return;
    }

    f.mode = el('select', {
      class: 'select',
      onChange: () => { state.monthMode = f.mode.value; drawFreq(); drawPreview(); },
    }, option('day', '日にちで決める（毎月◯日）', state.monthMode === 'day'),
    option('nth', '曜日で決める（第◯◯曜日）', state.monthMode === 'nth'));
    let detail;
    if (state.monthMode === 'nth') {
      const startDate = parseDate(f.start?.value) || now;
      f.nth = el('select', { class: 'select', onChange: drawPreview },
        ...NTHS.map(([value, label]) => option(value, label,
          value === Number(meeting?.nth ?? Math.min(4, Math.ceil(startDate.getDate() / 7))))));
      f.nthWeekday = el('select', { class: 'select', onChange: drawPreview },
        ...WEEKDAYS.map((label, index) => option(index, `${label}曜日`,
          index === Number(meeting?.nth_weekday ?? mondayIndex(startDate)))));
      detail = el('div', { style: { display: 'flex', gap: '6px' } }, f.nth, f.nthWeekday);
    } else {
      f.monthDay = el('input', {
        class: 'input', type: 'number', min: 1, max: 31, style: { maxWidth: '90px' },
        value: meeting?.month_day || (parseDate(f.start?.value) || now).getDate(),
      });
      f.monthDay.addEventListener('input', drawPreview);
      detail = el('div', { style: { display: 'flex', gap: '6px', alignItems: 'center' } },
        f.monthDay, el('span', { text: '日' }),
        el('span', { class: 'hint', text: '31 日などが無い月は、月末に行います' }));
    }
    fill(freqHost,
      el('div', { class: 'row' },
        el('div', { class: 'field', style: { flex: '0 0 130px' } },
          el('label', { text: '間隔' }), f.interval),
        el('div', { class: 'field' }, el('label', { text: '決め方' }), f.mode)),
      el('div', { class: 'field' }, el('label', { text: '開催日' }), detail));
  };

  const result = await openModal({
    title: meeting ? '定例を編集' : '定例を追加',
    build: () => {
      f.title = el('input', {
        class: 'input', maxlength: 200, value: meeting?.title || '',
        placeholder: '例）週次定例、ステアリングコミッティ',
      });
      // 置き場所。「開発フェーズの定例」のように、関係するタスクの下に並べられる
      f.parent = el('select', { class: 'select' },
        option('', '先頭の「定例」にまとめる', !meeting?.parent_id),
        ...placementCandidates(tasks).map((t) => option(t.id, `${'　'.repeat(t.depth)}${t.label}`,
          t.id === meeting?.parent_id)));
      if (meeting?.parent_id && ![...f.parent.options].some((o) => Number(o.value) === meeting.parent_id)) {
        // 候補に無いタスク（消えかけの行など）に置いてあっても、黙って外さない
        const current = tasks.find((t) => t.id === meeting.parent_id);
        f.parent.appendChild(option(meeting.parent_id, current?.title || `#${meeting.parent_id}`, true));
      }
      f.freq = el('div', { class: 'seg' }, ...[['weekly', '毎週・隔週'], ['monthly', '毎月'],
        ['dates', '日付を指定']]
        .map(([value, label]) => el('button', {
          type: 'button', class: state.freq === value ? 'active' : '',
          onClick: (event) => {
            if (state.freq !== value) state.interval = 1;
            state.freq = value;
            [...f.freq.children].forEach((b) => b.classList.toggle('active', b === event.currentTarget));
            if (!state.holidayTouched && DEFAULT_HOLIDAY[value]) {
              f.holiday.value = DEFAULT_HOLIDAY[value];
            }
            drawFreq();
            drawPreview();
          },
        }, label)));
      f.time = el('input', {
        class: 'input', maxlength: 20, value: meeting?.time_text || '',
        placeholder: '例）10:00', style: { maxWidth: '140px' },
      });
      f.holiday = el('select', { class: 'select' },
        ...HOLIDAY_RULES.map(([value, label]) => option(value, label,
          value === (meeting?.freq !== 'dates' && meeting?.holiday_rule
            ? meeting.holiday_rule : (DEFAULT_HOLIDAY[state.freq] || 'skip')))));
      f.holiday.addEventListener('change', () => { state.holidayTouched = true; drawPreview(); });
      f.start = el('input', {
        class: 'input', type: 'date',
        value: (meeting?.freq !== 'dates' && meeting?.start_on) || toISO(now),
      });
      f.end = el('input', { class: 'input', type: 'date', value: meeting?.end_on || '' });
      for (const input of [f.start, f.end]) input.addEventListener('change', drawPreview);
      f.time.addEventListener('input', drawPreview);
      const holidayField = el('div', { class: 'field' },
        el('label', { text: '休日に当たったら' }), f.holiday);
      const periodRow = el('div', { class: 'row' },
        el('div', { class: 'field' }, el('label', { text: '開始日' }), f.start),
        el('div', { class: 'field' }, el('label', { text: '終了日（空なら続ける）' }), f.end));
      ruleOnly.push(holidayField, periodRow);
      drawFreq();
      drawPreview();

      return el('div', {},
        el('div', { class: 'field' }, el('label', { text: '会議の名前 *' }), f.title),
        el('div', { class: 'field' }, el('label', { text: 'ガントで置く場所' }), f.parent,
          el('div', { class: 'hint',
            text: 'タスクを選ぶとその子として、見出しを選ぶとその区切りの先頭に並びます。'
              + '選んだタスクや見出しをたたむと一緒に隠れます。' })),
        el('div', { class: 'field' }, el('label', { text: '繰り返し' }), f.freq),
        freqHost,
        el('div', { class: 'row' },
          el('div', { class: 'field', style: { flex: '0 0 150px' } },
            el('label', { text: '時刻' }), f.time),
          holidayField),
        periodRow,
        preview,
        el('div', { class: 'hint', style: { marginTop: '8px' },
          text: '1 回ごとの中止や日にちの変更は、ガントの点を押して行えます。' }));
    },
    footer: (close) => [
      meeting ? el('button', {
        class: 'btn btn-danger', style: { marginRight: 'auto' },
        onClick: async () => {
          if (!await confirmDialog(
            `定例「${meeting.title}」を削除しますか？\n中止・日にち変更の記録も一緒に消えます。`,
            { danger: true, okLabel: '削除する' })) return;
          try {
            await api.del(`/api/meetings/${meeting.id}`);
            toast('削除しました', 'ok');
            close('deleted');
          } catch (error) { toast(error.message, 'error'); }
        },
      }, '削除') : null,
      el('button', { class: 'btn', onClick: () => close(null) }, 'キャンセル'),
      el('button', {
        class: 'btn btn-primary',
        onClick: async (event) => {
          const body = payload();
          if (!body.title) { toast('会議の名前を入れてください', 'error'); return; }
          if (body.freq === 'weekly' && !body.weekdays.length) {
            toast('曜日を 1 つ以上選んでください', 'error');
            return;
          }
          if (body.freq === 'dates' && !body.dates.length) {
            toast('日付を 1 つ以上選んでください', 'error');
            return;
          }
          const button = event.currentTarget;
          button.disabled = true;
          try {
            if (meeting) await api.patch(`/api/meetings/${meeting.id}`, body);
            else await api.post(`/api/projects/${project.id}/meetings`, body);
            toast(meeting ? '保存しました' : `定例「${body.title}」を追加しました`, 'ok');
            close(true);
          } catch (error) {
            toast(error.message, 'error');
            button.disabled = false;
          }
        },
      }, '保存'),
    ],
  });
  clearTimeout(previewTimer);
  return result;
}

/**
 * 1 回ぶんの操作。中止・日にち変更・元に戻す。変更したら true を返す。
 * occurrence.planned がその回の鍵（休日でずらしたあとの予定日）。
 */
export async function openOccurrenceDialog(meeting, occurrence) {
  const planned = occurrence.planned;
  const f = {};
  const statusText = {
    normal: '予定どおり',
    cancelled: '中止',
    moved: `${dayLabel(planned)} から ${dayLabel(occurrence.date)} に変更`,
  }[occurrence.status];

  const send = async (close, button, body, message) => {
    button.disabled = true;
    try {
      await api.put(`/api/meetings/${meeting.id}/exceptions/${planned}`, body);
      toast(message, 'ok');
      close(true);
    } catch (error) {
      toast(error.message, 'error');
      button.disabled = false;
    }
  };

  return openModal({
    title: `${meeting.title} — ${dayLabel(occurrence.date)}`,
    build: () => {
      f.note = el('input', {
        class: 'input', maxlength: 200, value: occurrence.note || '',
        placeholder: '例）出張のため、全社行事と重なるため',
      });
      f.moveTo = el('input', {
        class: 'input', type: 'date', style: { maxWidth: '170px' },
        value: occurrence.status === 'moved' ? occurrence.date : planned,
      });
      return el('div', {},
        el('p', { class: 'page-sub' },
          `${meeting.summary} の回です。いまの状態: `, el('strong', { text: statusText })),
        occurrence.shifted_from
          ? el('div', { class: 'hint', style: { marginBottom: '10px' },
            text: `本来は ${dayLabel(occurrence.shifted_from)} ですが、休日のためこの日にずらしています。` })
          : null,
        meeting.project_id && occurrence.status !== 'cancelled'
          && (store.project(meeting.project_id)?.tabs || ['decisions']).includes('decisions')
          ? el('div', { class: 'meeting-decision' },
            el('span', { class: 'hint', text: 'この回で決まったことを、意思決定ログに残せます' }),
            el('button', {
              type: 'button', class: 'btn btn-sm',
              onClick: async () => {
                const list = await api.get(`/api/projects/${meeting.project_id}/decisions`);
                const { openDecisionForm } = await import('./decisionForm.js');
                const saved = await openDecisionForm({
                  projectId: meeting.project_id, decisions: list.decisions,
                  prefill: {
                    title: '', status: 'decided', decided_on: occurrence.date,
                    meeting_id: meeting.id, meeting_on: occurrence.date, meeting_title: meeting.title,
                  },
                });
                if (saved) toast(`D-${saved.seq} を記録しました（決定タブで見られます）`, 'ok');
              },
            }, '⚖️ 決定として記録'))
          : null,
        el('div', { class: 'field' }, el('label', { text: 'メモ（理由など）' }), f.note),
        el('div', { class: 'field' },
          el('label', { text: '日にちを変えるなら' }),
          el('div', { style: { display: 'flex', gap: '8px', alignItems: 'center' } },
            f.moveTo, el('span', { class: 'hint', text: 'に変える' }))));
    },
    footer: (close) => [
      occurrence.status !== 'normal' ? el('button', {
        class: 'btn', style: { marginRight: 'auto' },
        title: '中止・日にち変更を取り消して、決まりどおりに戻します',
        onClick: async (event) => {
          const button = event.currentTarget;
          button.disabled = true;
          try {
            await api.del(`/api/meetings/${meeting.id}/exceptions/${planned}`);
            toast(`${dayLabel(planned)} を予定どおりに戻しました`, 'ok');
            close(true);
          } catch (error) {
            toast(error.message, 'error');
            button.disabled = false;
          }
        },
      }, '予定どおりに戻す') : null,
      el('button', {
        class: 'btn',
        onClick: (event) => {
          const target = f.moveTo.value;
          if (!target) { toast('変更先の日付を選んでください', 'error'); return; }
          if (target === planned) {
            toast('元の日と同じです。予定どおりに戻すなら左のボタンを押してください', 'error');
            return;
          }
          send(close, event.currentTarget,
            { action: 'move', moved_to: target, note: f.note.value },
            `${dayLabel(planned)} の回を ${dayLabel(target)} に変えました`);
        },
      }, '日にちを変える'),
      el('button', {
        class: 'btn btn-danger',
        onClick: (event) => send(close, event.currentTarget,
          { action: 'cancel', note: f.note.value },
          `${dayLabel(planned)} の回を中止にしました`),
      }, occurrence.status === 'cancelled' ? 'メモを保存' : 'この回を中止'),
    ],
  });
}

/** 点の吹き出しに出す文。 */
export function occurrenceTitle(meeting, occurrence, projectName = '') {
  const lines = [
    `${projectName ? `${projectName} / ` : ''}${meeting.title}`,
    `${dayLabel(occurrence.date)}${meeting.time_text ? ` ${meeting.time_text}` : ''}`,
  ];
  if (occurrence.status === 'cancelled') lines.push('中止');
  if (occurrence.status === 'moved') lines.push(`${dayLabel(occurrence.planned)} から変更`);
  if (occurrence.shifted_from) lines.push(`${dayLabel(occurrence.shifted_from)} が休日のためずらし`);
  if (occurrence.note) lines.push(occurrence.note);
  return lines.join('\n');
}

